airbyte-cdk/python/airbyte_cdk/sources/concurrent_source/concurrent_stream_processor.py

```diff
@@ -0,0 +1,189 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+import logging
+from typing import Dict, Iterable, List, Optional, Set
+
+from airbyte_cdk.models import AirbyteMessage, AirbyteStreamStatus
+from airbyte_cdk.models import Type as MessageType
+from airbyte_cdk.sources.concurrent_source.partition_generation_completed_sentinel import PartitionGenerationCompletedSentinel
+from airbyte_cdk.sources.concurrent_source.thread_pool_manager import ThreadPoolManager
+from airbyte_cdk.sources.message import MessageRepository
+from airbyte_cdk.sources.streams.concurrent.abstract_stream import AbstractStream
+from airbyte_cdk.sources.streams.concurrent.partition_enqueuer import PartitionEnqueuer
+from airbyte_cdk.sources.streams.concurrent.partition_reader import PartitionReader
+from airbyte_cdk.sources.streams.concurrent.partitions.partition import Partition
+from airbyte_cdk.sources.streams.concurrent.partitions.record import Record
+from airbyte_cdk.sources.streams.concurrent.partitions.types import PartitionCompleteSentinel
+from airbyte_cdk.sources.utils.record_helper import stream_data_to_airbyte_message
+from airbyte_cdk.sources.utils.slice_logger import SliceLogger
+from airbyte_cdk.utils.stream_status_utils import as_airbyte_message as stream_status_as_airbyte_message
+
+
+class ConcurrentStreamProcessor:
+    def __init__(
+        self,
+        stream_instances_to_read_from: List[AbstractStream],
+        partition_enqueuer: PartitionEnqueuer,
+        thread_pool_manager: ThreadPoolManager,
+        logger: logging.Logger,
+        slice_logger: SliceLogger,
+        message_repository: MessageRepository,
+        partition_reader: PartitionReader,
+    ):
+        """
+        This class is responsible for handling items from a concurrent stream read process.
+        :param stream_instances_to_read_from: List of streams to read from
+        :param partition_enqueuer: PartitionEnqueuer instance
+        :param thread_pool_manager: ThreadPoolManager instance
+        :param logger: Logger instance
+        :param slice_logger: SliceLogger instance
+        :param message_repository: MessageRepository instance
+        :param partition_reader: PartitionReader instance
+        """
+        self._stream_name_to_instance = {s.name: s for s in stream_instances_to_read_from}
+        self._record_counter = {}
+        self._streams_to_partitions: Dict[str, Set[Partition]] = {}
+        for stream in stream_instances_to_read_from:
+            self._streams_to_partitions[stream.name] = set()
+            self._record_counter[stream.name] = 0
+        self._thread_pool_manager = thread_pool_manager
+        self._partition_enqueuer = partition_enqueuer
+        self._stream_instances_to_read_from = stream_instances_to_read_from
```

Comment:
I think the naming is good but I still had to check the usage to make sure I understand the scope. Can we make this more explicit? `_streams_to_start_partition_generation` seems clumsy but feels a little more explicit to me

The other option I would see if so have a map of streams with the step they are at. The reason I'm suggesting this is that is seems like `_streams_currently_generating_partitions` also describe a step a stream is at and maybe `_stream_current_steps: Mapping[str, ProcessingStep]` where:
```
class ProcessingStep(Enum):
  NOT_STARTED,
  PROCESSING,
  DONE
```

What I don't like with this solution is that as the code is executed concurrently, there can be multiple steps going on at the same time (for example if we wanted to add "partition generation" and "records generation") so it seems hard to evolve the object